uart_sam0: don't consume CONFIG_UART_ASYNC_API directly.

Flameeyes opened the pull request:
uart_sam0: don't consume CONFIG_UART_ASYNC_API directly.


There are situations in which the async API for uart can be provided
by another driver (case in point, uart_rtt), and thus there is no valid
DMA controller for the uart_sam0 driver to talk with.

By separating the configuration, there's no need to exclude samd20-based
boards (that have no DMA peripheral) from the uart_async_api tests.

Signed-off-by: Diego Elio Pettenò <flameeyes@meta.com>

Flameeyes commented:
Note: this is really only a bugfix for samd20 boards, since they don't support DMA so they shouldn't enable the UART_SAM0_ASYNC at all.

But for samd21 boards this is just a workaround to keep the code disabled, since trying to enable it fails to build (#61198).

Flameeyes commented:
Rebased to solve conflict with #61140

nandojve commented:
Hi @Flameeyes ,

I was looking on this more in deep and attached a diff that I would like to add on this PR. I'm not an expert on Kconfig but this was what I understood:

1. The meaning of SERIAL_SUPPORT_ASYNC is to tell that driver supports the ASYNC API and not enable it. Because of that I think the `select SERIAL_SUPPORT_ASYNC if DT_HAS_ATMEL_SAM0_DMAC_ENABLED` is still valid, since it only informs that user can use serial on async mode.

2. The UART_SAM0_ASYNC that you proposed should be only available to the user/application define the UART_ASYNC_API=y and DMA_SAM0=y.

Do you think my assumptions make sense?

patch: [fix_sam0_uart_and_test_configs.txt](https://github.com/zephyrproject-rtos/zephyr/files/12363860/fix_sam0_uart_and_test_configs.txt)

Flameeyes committed:
uart_sam0: don't consume CONFIG_UART_ASYNC_API directly.

There are situations in which the async API for uart can be provided
by another driver (case in point, uart_rtt), and thus there is no valid
DMA controller for the uart_sam0 driver to talk with.

By separating the configuration, there's no need to exclude samd20-based
boards (that have no DMA peripheral) from the uart_async_api tests.

Signed-off-by: Diego Elio Pettenò <flameeyes@meta.com>
Co-authored-by: Gerson Fernando Budke <nandojve@gmail.com>

Flameeyes commented:
That makes sense to me — I don't expect there's much of a chance of someone wanting RTT as async and SAM0 as not.

And nice to be able to cover a lot more boards in tests 🥳

nandojve commented:
> That makes sense to me — I don't expect there's much of a chance of someone wanting RTT as async and SAM0 as not.
> 
> And nice to be able to cover a lot more boards in tests 🥳

I was wondering if you would mind to create another commit with the content that I sent to you and keep me in as Co-authored-by: Gerson Fernando Budke [nandojve@gmail.com](mailto:nandojve@gmail.com). You don't need add in this new commit the content related to Kconfig.sam0 since we both worked on that.

Flameeyes commented:
> I was wondering if you would mind to create another commit with the content that I sent to you and keep me in as Co-authored-by: Gerson Fernando Budke [nandojve@gmail.com](mailto:nandojve@gmail.com). You don't need add in this new commit the content related to Kconfig.sam0 since we both worked on that.

Ah sure thing, let me make this two commits, since they can be made into two "cleanly" separate ones.

nandojve committed:
uart_sam0: extend async serial tests to more atsamxxx_xpro boards.

This includes new overlays for boards supporting DMA, and one more test
to verify M0 boards both with and without DMA can be built when
requesting CONFIG_UART_ASYNC_API.

Signed-off-by: Diego Elio Pettenò <flameeyes@meta.com>
Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
The fail at Unit Test is unrelated and it seems a bug on modem code. It has been addressed at modem channel.
